File: blitzwing-executor/src/main/scala/com/ebay/hadoop/arrow/executor/ArrowReferenceManager.scala
package com.ebay.hadoop.arrow.executor

import java.util.concurrent.atomic.AtomicInteger

import io.netty.buffer.ArrowBuf
import org.apache.arrow.memory.{BufferAllocator, OwnershipTransferResult, ReferenceManager}

class ArrowReferenceManager(private val memory: ArrowBufferWrapper) extends ReferenceManager {
  private val bufRefCnt = new AtomicInteger(0)

  override def getRefCount: Int = {
    bufRefCnt.get()
  }

  override def release(): Boolean = release(1)

  override def release(decrement: Int): Boolean = {
    require(decrement >= 1, "ref count decrement should be greater than or equal to 1")
    // decrement the ref count
    var refCnt = 0
    this synchronized {
      refCnt = bufRefCnt.addAndGet(-decrement)
      if (refCnt == 0) { // refcount of this reference manager has dropped to 0
        // release the underlying memory
        memory.close()
      }
    }

    // the new ref count should be >= 0
    require(refCnt >= 0, "RefCnt has gone negative")
    refCnt == 0
  }

  override def retain(): Unit = retain(1)

  override def retain(increment: Int): Unit = {
    require(increment >= 1, s"retain argument($increment) is not positive")
    bufRefCnt.addAndGet(increment)
  }

  override def retain(srcBuffer: ArrowBuf, targetAllocator: BufferAllocator): ArrowBuf = {
    retain()
    srcBuffer
  }

  override def deriveBuffer(sourceBuffer: ArrowBuf, index: Int, length: Int): ArrowBuf = ???

  override def transferOwnership(sourceBuffer: ArrowBuf, targetAllocator: BufferAllocator): OwnershipTransferResult = {
    throw new UnsupportedOperationException("transferOwnership")
  }

  override def getAllocator: BufferAllocator = null

  override def getSize: Int = memory.length.toInt

  override def getAccountedSize: Int = 0
}

File: blitzwing-common/src/main/java/com/ebay/hadoop/blitzwing/memory/ArrowReferenceManager.java
package com.ebay.hadoop.blitzwing.memory;

import io.netty.buffer.ArrowBuf;
import java.util.Objects;
import java.util.concurrent.atomic.AtomicInteger;
import org.apache.arrow.memory.BufferAllocator;
import org.apache.arrow.memory.OwnershipTransferResult;
import org.apache.arrow.memory.ReferenceManager;
import org.apache.arrow.util.Preconditions;

public class ArrowReferenceManager implements ReferenceManager {
  private final ArrowBufferWrapper buffer;
  private final AtomicInteger bufferRefCount = new AtomicInteger(0);

  public ArrowReferenceManager(ArrowBufferWrapper buffer) {
    this.buffer = buffer;
  }


  @Override
  public int getRefCount() {
    return bufferRefCount.get();
  }

  @Override
  public boolean release() {
    return release(1);
  }

  @Override
  public boolean release(int decrement) {
    Preconditions.checkArgument(decrement >= 1, "ref count decrement should be greater than or equal to 1");

    synchronized (this) {
      int refCount = bufferRefCount.addAndGet(-decrement);
      if (refCount == 0) {
        buffer.close();
      }

      if (refCount < 0) {
        throw new IllegalStateException("Ref count has gone negative: " + refCount);
      }
      return refCount == 0;
    }
  }

  @Override
  public void retain() {
    retain(1);
  }

  @Override
  public void retain(int increment) {
    Preconditions.checkArgument(increment >= 1, "Retain argument is not positive");
    bufferRefCount.addAndGet(increment);
  }

  @Override
  public ArrowBuf retain(ArrowBuf srcBuffer, BufferAllocator targetAllocator) {
    retain();
    return srcBuffer;
  }

  @Override
  public ArrowBuf deriveBuffer(ArrowBuf sourceBuffer, int index, int length) {
    throw new UnsupportedOperationException("Derive buffer is not supported!");
  }

  @Override
  public OwnershipTransferResult transferOwnership(ArrowBuf sourceBuffer,
      BufferAllocator targetAllocator) {
    throw new UnsupportedOperationException("transferOwnership is not supported!");
  }

  @Override
  public BufferAllocator getAllocator() {
    return null;
  }

  @Override
  public int getSize() {
    return buffer.getLength();
  }

  @Override
  public int getAccountedSize() {
    return 0;
  }
}
